### PR TITLE
Secure webhook calls

### PR DESCRIPTION
**Pull request by stanleydunne**
Verify all the incoming webhook calls, to ensure they are coming from a trusted source.
@bhumit can you review and merge.

### CODE REVIEW

File: src/main/java/com/ibm/watsonwork/controller/WatsonWorkController.java [hunk 1]
```diff
@@ -97,8 +98,28 @@ public String oauthCallback(@RequestParam(CODE_VALUE) String code, @RequestParam
     }
 
     @PostMapping(value = "/webhook", produces = MediaType.APPLICATION_JSON_VALUE)
-    public ResponseEntity webhookCallback(@RequestHeader(X_OUTBOUND_TOKEN) String outboundToken, @RequestBody WebhookEvent webhookEvent) {
-        if (VERIFICATION.equalsIgnoreCase(webhookEvent.getType()) && authService.isValidVerificationRequest(webhookEvent, outboundToken)) {
+    public ResponseEntity webhookCallback(@RequestHeader(X_OUTBOUND_TOKEN) String outboundToken, @RequestBody String webhookEventString) {
+		//log.info("Raw webHookEvent : \"" + webhookEventString + "\"" );
+
+		if(!verifyWebHookRequest(webhookEventString, outboundToken))
+		{
```
Comment (bhumit) [inline src/main/java/com/ibm/watsonwork/controller/WatsonWorkController.java]: could we conform to the codestyle currently across the project?

File: src/main/java/com/ibm/watsonwork/controller/WatsonWorkController.java [hunk 2]
```diff
@@ -97,8 +98,28 @@ public String oauthCallback(@RequestParam(CODE_VALUE) String code, @RequestParam
     }
 
     @PostMapping(value = "/webhook", produces = MediaType.APPLICATION_JSON_VALUE)
-    public ResponseEntity webhookCallback(@RequestHeader(X_OUTBOUND_TOKEN) String outboundToken, @RequestBody WebhookEvent webhookEvent) {
-        if (VERIFICATION.equalsIgnoreCase(webhookEvent.getType()) && authService.isValidVerificationRequest(webhookEvent, outboundToken)) {
+    public ResponseEntity webhookCallback(@RequestHeader(X_OUTBOUND_TOKEN) String outboundToken, @RequestBody String webhookEventString) {
+		//log.info("Raw webHookEvent : \"" + webhookEventString + "\"" );
```
Comment (bhumit) [inline src/main/java/com/ibm/watsonwork/controller/WatsonWorkController.java]: remove?

File: src/main/java/com/ibm/watsonwork/controller/WatsonWorkController.java [hunk 3]
```diff
@@ -97,8 +98,28 @@ public String oauthCallback(@RequestParam(CODE_VALUE) String code, @RequestParam
...
+
+		if(!verifyWebHookRequest(webhookEventString, outboundToken))
+		{
+			log.info("In the WebHook request failed security verifcation:" + webhookEventString);
+	        return ResponseEntity.status(403).build();     // return a forbidden status
```
Comment (bhumit) [inline src/main/java/com/ibm/watsonwork/controller/WatsonWorkController.java]: `ResponseEntity.status(HttpStatus.FORBIDDEN).build()` ?

File: src/main/java/com/ibm/watsonwork/controller/WatsonWorkController.java [hunk 4]
```diff
@@ -97,8 +98,28 @@ public String oauthCallback(@RequestParam(CODE_VALUE) String code, @RequestParam
...
+	    ObjectMapper objectMapper = new ObjectMapper();          
+        WebhookEvent webhookEvent;
+        try
+		{
+			webhookEvent = objectMapper.readValue(webhookEventString, WebhookEvent.class);
+		}
+		catch(Exception e)
+		{
+			log.info("In the WebHook body string not parsed:" + webhookEventString);
+	        return ResponseEntity.status(400).build();     // return a bad request status
```
Comment (bhumit) [inline src/main/java/com/ibm/watsonwork/controller/WatsonWorkController.java]: ResponseEntity.status(HttpStatus.BAD_REQUEST).build()

File: src/main/java/com/ibm/watsonwork/controller/WatsonWorkController.java [hunk 5]
```diff
@@ -97,8 +98,28 @@ public String oauthCallback(@RequestParam(CODE_VALUE) String code, @RequestParam
...
+    public ResponseEntity webhookCallback(@RequestHeader(X_OUTBOUND_TOKEN) String outboundToken, @RequestBody String webhookEventString) {
+		//log.info("Raw webHookEvent : \"" + webhookEventString + "\"" );
+
+		if(!verifyWebHookRequest(webhookEventString, outboundToken))
+		{
+			log.info("In the WebHook request failed security verifcation:" + webhookEventString);
+	        return ResponseEntity.status(403).build();     // return a forbidden status
+		}
+        
+	    ObjectMapper objectMapper = new ObjectMapper();
```
Comment (bhumit) [inline src/main/java/com/ibm/watsonwork/controller/WatsonWorkController.java]: do you need to instantiate the `ObjectMapper`?

File: src/main/java/com/ibm/watsonwork/controller/WatsonWorkController.java [hunk 6]
```diff
@@ -97,8 +98,28 @@ public String oauthCallback(@RequestParam(CODE_VALUE) String code, @RequestParam
...
+        
+	    ObjectMapper objectMapper = new ObjectMapper();          
+        WebhookEvent webhookEvent;
+        try
+		{
+			webhookEvent = objectMapper.readValue(webhookEventString, WebhookEvent.class);
+		}
+		catch(Exception e)
+		{
+			log.info("In the WebHook body string not parsed:" + webhookEventString);
```
Comment (bhumit) [inline src/main/java/com/ibm/watsonwork/controller/WatsonWorkController.java]: perhaps `Failed to parse webhook event`

File: src/main/java/com/ibm/watsonwork/controller/WatsonWorkController.java [hunk 7]
```diff
@@ -129,20 +150,43 @@ private void processWebhook(WebhookEvent webhookEvent) {
...
                 .body(responseBody);
     }
 
+
+    private boolean verifyWebHookRequest(String body, String header)
+    {
+		try
+		{
+			String verification = authService.createVerificationHeader(body);
+			if(verification.equals(header))
```
Comment (bhumit) [inline src/main/java/com/ibm/watsonwork/controller/WatsonWorkController.java]: `return verification.equals(header)`

**Review by bhumit — CHANGES_REQUESTED** [review]

**Comment by stanleydunne** [comment]
@bhumit changes made per review comments.
I do have to have an instance of the ObjectMapper, so that stays.
So wish WW didn't use the ugly curly brace style :-)

File: src/main/java/com/ibm/watsonwork/controller/WatsonWorkController.java [hunk 8]
```diff
@@ -170,16 +165,10 @@ private ResponseEntity buildVerificationResponse(WebhookEvent webhookEvent) {
 
     private boolean verifyWebHookRequest(String body, String header)
     {
```
Comment (bhumit) [inline src/main/java/com/ibm/watsonwork/controller/WatsonWorkController.java]: missed one?

**Comment by bhumit** [comment]
Can you clarify what you mean by ugly brace style? We are using Oracle's code conventions (http://www.oracle.com/technetwork/java/codeconventions-141270.html#381)

**Comment by stanleydunne** [comment]
Hopefully all okay now.
It's a personal thing; I have a huge preference for braces on their own lines, in my opinion it makes code much easier to read. I never liked the original C K&R styling.

**Review by bhumit — APPROVED** [review]

**Pull request merged**